README
What is GnoTime?
================
The Gnome Time Tracker is a to-do list/diary/journal tool that can track
the amount of time spent on projects, and, among other things, generate
reports and invoices based on that time. It can be used to keep shopping
lists, organize ideas, track bug reports, keep a diary of activities,
provide weekly status reports to management, and even works as a consultant
billing system.

***IMPORTANT*** *Get the latest code from here:*
https://github.com/markuspg/gnotime/tree/modernize3
It will compile & run on modern (2023-era) Linux systems!

That code will be soon be merged into
https://github.com/GnoTime/gnotime
so be sure to check that, too.

Everything below is stale and has not been updated since 2013.

HomePage
--------
 * http://gttr.sourceforge.net
 * http://www.linas.org/linux/gtt/gtt.html

Features
--------
 * TODO Lists
 * Diary/Journal
 * Multiple task timers
 * Billing subsystem
 * (Configurable) HTML Reports

Please see the [GnoTime web page](http://gttr.sourceforge.net) for a
detailed description of these features.

Status
------
The newest and latest version can be found in the "modernize3" branch of
the @markuspg fork:

https://github.com/markuspg/gnotime/tree/modernize3

The above will compile and run on present-era (2023) Linux systems.

That code will soon be merged into
https://github.com/GnoTime/gnotime
and so be sure to check that repo, too.

Note that the above repos are NOT forked from this repo! So if you are
just looking at forks, you won't find them.


Obsolete commentary from 2013
-----------------------------
Gnotime is written for Gnome-1/Gnome-2. It needs to be ported to
Gnome3 to be buildable on present-day systems.

Porting to gnome3 is probably not that hard(?) There is one sticky
point: the main display panel uses the very old gnome-1 `gtkctree`
widget. The gnome2 `treeview` widget was a horrible, terrible
replacement for `gtkctree`, and was never used (was unusable).
It's not clear if gnome3 has any suitable replacement.  See the
notes in `src/ctree-gnome2.c` for details. Basically, the main
window depends heavily on keyboard/mouse navigation, which the
`treeview` widget completely failed to support.

Put it differently: if the main window isn't nice and pleasant to
use, then nothing else matters.  Having a good user experience
working with the main window is more important than anything else.

If you want to see how it works, install an LXC/LXD container
(or even docker) with a circa 2012 or 2014 Ubuntu or Debian system.
It should build cleanly, and run bug-free w/o issues.

Building
--------

### Required packages:
```
guile-3.0-dev
libgtk2.0-dev
libglade2-dev
libgconf2-dev
libxss-dev
libxml2-dev
```

Developers and maintainers also need:
```
glade-gnome
```

Where have these wandered off to ???
```
libgnome2-dev
libgnomevfs2-dev
libgnomeui2-dev
libgtkhtml-dev
```

### QOF Query Object Framework
One pre-requiste to building this is the `qof` package.
It is not distributed by distros, by default.

QOF on github:
 * https://github.com/codehelp/qof

QOF Documentation:
 * http://qof.sourceforge.net/doxy/main.html

It's not hard to build:
```
git clone https://github.com/codehelp/qof
./autogen.sh
./configure
make
sudo make install
```


### Build steps:
```
./autogen.sh --no-configure
mkdir build; cd build
../configure
```

Authors
-------
This program was originally written by Eckehard Berns <eb@berns.prima.de>,
but has been greatly expanded by Linas Vepstas ~~<linas@linas.org>~~
<linasvepstas@gmail.com>.

Thanks go out to many people who e-mailed me with suggestions and
bug fixes.  See the "about" window in the app for more details.


Copyright
---------
 * GnoTime - the Gnome Time Tracker
 * Copyright (C) 1997,98 Eckehard Berns
 * Copyright (C) 2001,2002,2003,2004 Linas Vepstas <linas@linas.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   For more details see the file [COPYING](COPYING).


NEWS


                 GnoTime Application News
                 ========================

GnoTime, The Gnome Time Tracker, allows you to organize ideas,
measure the time spent on to-do list items and generate invoices.

The latest releases continues to focus on usability and convenience
features, making GnoTime easier to use and thus more powerful.

New in the 2.3.0 Release:
-------------------------
The main changes in this release are:
* a new implementation of the projects tree using the GtkTreeView
  component which fixes a lot of the bugs regarding the projects
  tree manipulation and prepares the way for many customizations
  in future releases.
* an initial implementation of a tray icon that displays timer status
  and allows the user to toggle the timer by clicking on it.
* a rewrite of the idle detection using the XScreenSaver extension and
  of the timer system to make it more system power conservation
  friendly.
* port of the report system to use guile version 1.8.
* many changes in the configure script, including porperly checking for
  guile and gconf and version of available DBUS.
* removal of some deprecated APIs usage.


In addition, many bugs have been fixed, including:
* broken behavior of drag-and-drop (#1856268)
* compiles but crashes on startup in fedora core 5 (#1802501)
* activity journal truncates seconds (#1706365)
* green bar flickers when value changes (#1171204)
* dragging project to make sub-project will not scroll (#831808)

New in the 2.2.3 Release:
-------------------------
This release brings better internationalization in reports with locale
based formating of times and monetary values. The newly added support
for DBus messages allows users to start/stop the timer and save/reload
the data file from the command line using the gnotime-remote
script. Also, an update user's manual covers most of the features
available in this version. In addition, many bugs have been fixed,
including:

-- Crash when saving to full filesystem (#1548248)
-- Day of century calculation error causing skipped day in reports
-- Conflicting keybindings (#1639477)
-- Invoice report doesn't work in locales other than C
-- Hidden question dialog when overwriting file during report saving
   (http://bugs.debian.org/422023)

New in the 2.2.0 Release:
-------------------------
This release sports a brand-new look-n-feel for the reports,
together with improved URL support.  The reports now use CSS style 
sheets, and can contain embedded images.  A 'publish' function 
using GnomeVFS makes it easier to copy reports to other systems, 
using ssh, ftp, http or any other supported method.  In addition, 
the following bugs were fixed:

-- Crashes when 'help' button clicked in certain dialogs.
-- Corruption of 'reports' menu, followed by crash.
-- Fix flyover help to actually work.
-- Add missing 'copy' to interval edit popup menu.

New in the 2.1.9 Release:
-------------------------
This release fixes a number of bugs, the most serious of which were:
-- Keyboard events were not resetting the timeout
-- Fix Help system so that help actually starts up.
-- Fix old, multiple scheme bugs affecting billing reports.
-- Provides 'infinite undo' for cut projects.
-- Keyboard cut (ctrl-x) fixed so that it cuts only if project 
   window has focus.
-- Can now specify day-of-week for the start of a new week.
   Also, end-of-day need not be midnight - its now configurable.
-- Remove obsolete menu entries (clear daily counters)
-- Fix flyover help on HTML reports
-- Fix order of yesterday/today columns.
-- Improve checks for Guile during builds
-- New whiz-bang SQL for reports.


New in the 2.1.8 release:
-------------------------
This release adds support for simple report queries.  In particular,
reports can now be generated for all tasks that occured between a
pair of dates;  an extremely general query system is now in place,
although it is not used except to perform this simple date query.

Toshio Kuratomi <toshio@tiki-lounge.com> provided a major 
cleanup of the documentation subsystem.  Goedson Teixeira Paixao
provided a new pt_BR translation.

Bug fixes:
-- avoid crash of 'yelp' help browser due to SIGCHILD zombie reap.
-- numerous other minor crashes fixed.


New in the 2.1.7 release:
-------------------------
Version 2.1.7 fixes some internationalization bugs, and ports the 
configuration system to GConf2. The 'custom reports' menu is now 
editable.  


New in the 2.1.6 release:
-------------------------
The "Daily Report" will show not only the total hours worked,
day by day, but will also show a bullited list of tasks worked
that day.  The main window display can now show total hours worked
for previous day and previous week.  Sub-projects can now be fetched
via scheme, allowing them to be included in reports.  The scheme
subsystem now includes an elementary type system, allowing more
complex structures to be built.  A half-dozen different crashes
have been fixed.


New in the 2.1.5 release:
-------------------------
The main window has been reformulated in a big way:  It
now includes sub-panels for editing project and diary notes.
This 'notes area' makes it far more convenient to make
extensive notes and 'blog' your way through a set of projects.

Also fixes bugs with backup timestamps, also with cut-n-paste
of project trees.


New in the 2.1.4 release:
-------------------------
The scheme interface re-design has been finished.
This means that the HTML reports are far, far more
configurable, and can be made to look much prettier.
The old, nasty table-formatting features have been removed.
It also means that tab-delimited export now works cleanly,
and other export formats can be easily added.


New in the 2.1.3 release:
-------------------------
The time credited to a project after the keyboard inactivity 
timer has tripped is now very easily adjustable.  Thus, if
the timer tripped because you were on the phone, you can now
trivially credit the project with any time, from zero to 

src/app.c
/*   GTimeTracker - a time tracker
 *   Copyright (C) 1997,98 Eckehard Berns
 *   Copyright (C) 2001,2002,2003,2004 Linas Vepstas <linas@linas.org>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
#include "config.h"

#include <gnome.h>
#include <sched.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include <qof.h>

#include "active-dialog.h"
#include "app.h"
#include "cur-proj.h"
#include "gtt.h"
#include "log.h"
#include "menucmd.h"
#include "menus.h"
#include "notes-area.h"
#include "prefs.h"
#include "props-proj.h"
#include "timer.h"
#include "toolbar.h"
#include "util.h"
#include "projects-tree.h"
#include "proj.h"

/* XXX Most of the globals below should be placed into a single
 * application-wide top-level structure, rather than being allowed
 * to be globals. But, for now, its OK ...
 */
GttProject *cur_proj = NULL;

GttProjectsTree *projects_tree = NULL;
NotesArea *global_na = NULL;
GtkWidget *app_window = NULL;
GtkWidget *status_bar = NULL;


static GtkLabel *status_project = NULL;
static GtkLabel *status_day_time = NULL;
static GtkWidget *status_timer = NULL;

char *config_shell_start = NULL;
char *config_shell_stop = NULL;

gboolean geom_place_override = FALSE;
gboolean geom_size_override = FALSE;


extern GttActiveDialog *act;

static void
projects_tree_row_activated (GtkTreeView *tree_view, GtkTreePath *path, GtkTreeViewColumn *column, gpointer user_data)
{
	GttProjectsTree *gpt = GTT_PROJECTS_TREE (tree_view);
	GttProject *prj = gtt_projects_tree_get_selected_project (gpt);
	if (cur_proj == prj)
	{
		cur_proj_set (NULL);
	}
	else
	{
		cur_proj_set (prj);
	}
}

typedef void (*sort_function) (GttProjectList *);

static void
column_clicked (GtkTreeViewColumn *column, gpointer user_data)
{
	sort_function f = (sort_function) user_data;

	f (global_plist);
	gchar *expander_state = gtt_projects_tree_get_expander_state (projects_tree);
	gtt_projects_tree_populate (projects_tree, gtt_project_list_get_list (global_plist), TRUE);
	gtt_projects_tree_set_expander_state (projects_tree, expander_state);
	gtt_projects_tree_set_sorted_column (projects_tree, column);

}

static void
projects_tree_columns_setup_done (GttProjectsTree *projects_tree, gpointer user_data)
{

	GtkTreeViewColumn *column = NULL;

	column = gtt_projects_tree_get_column_by_name (projects_tree, "time_ever");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_ever);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "time_year");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_year);

	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "time_month");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_month);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "time_week");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_week);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "time_lastweek");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_lastweek);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "time_yesterday");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_yesterday);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "time_today");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_day);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "time_task");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_current);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "title");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_title);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "description");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_desc);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "estimated_start");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_start);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "estimated_end");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_end);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "due_date");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_due);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "sizing");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_sizing);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "percent_done");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_percent);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "urgency");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_urgency);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "importance");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_importance);
	}

	column = gtt_projects_tree_get_column_by_name (projects_tree, "status");
	if (column)
	{
		g_signal_connect (column, "clicked", G_CALLBACK (column_clicked), project_list_sort_status);
	}
}

/* ============================================================= */

void
update_status_bar(void)
{
	char day_total_str[25];
	char *s;

	if (!status_bar) return;

	/* Make the little clock item appear/disappear
	 * when the project is started/stopped */
	if (status_timer)
	{
		if (timer_is_running())
			gtk_widget_show(status_timer);
		else
			gtk_widget_hide(status_timer);
	}

	/* update timestamp */
	xxxqof_print_hours_elapsed_buff (day_total_str, 25,
	                                 gtt_project_list_total_secs_day(),
	                                 config_show_secs);

	if (0 != strcmp(day_total_str, gtk_label_get_text(status_day_time)))
	{
		gtk_label_set_text(status_day_time, day_total_str);
	}

	/* Display the project title */
	if (cur_proj)
	{
		s = g_strdup_printf ("%s - %s",
	                       gtt_project_get_title(cur_proj),
	                       gtt_project_get_desc(cur_proj));
	}
	else
	{
		s = g_strdup(_("Timer is not running"));
	}

	if (0 != strcmp(s, gtk_label_get_text(status_project)))
	{
		gtk_label_set_text(status_project, s);
	}
	g_free(s);
}


/* ============================================================= */
/* Handle shell commands */

void
do_run_shell_command (const char * str)
{
	pid_t pid;
	char *shell_path;
	int rc;
	struct stat shat;

	/* XXX This whole thing needs to be reviewewed for security */

	 /* Provide minimal security by using only system shells */
	shell_path = "/bin/sh";
	rc = stat (shell_path, &shat);
	if ((0 == rc) && S_ISREG (shat.st_mode) && (S_IXUSR & shat.st_mode))
	{
		goto do_run_shell;
	}

	shell_path = "/usr/bin/sh";
	rc = stat (shell_path, &shat);
	if ((0 == rc) && S_ISREG (shat.st_mode) && (S_IXUSR & shat.st_mode))
	{
		goto do_run_shell;
	}
	g_warning("%s: %d: do_run_shell_command: can't find shell\n", __FILE__, __LINE__);
	return;

do_run_shell:
	pid = fork();
	if (pid == 0)
	{
		execlp(shell_path, shell_path, "-c", str, NULL);
		g_warning("%s: %d: do_run_shell_command: couldn't exec\n", __FILE__, __LINE__);
		exit(1);
	}
	if (pid < 0)
	{
		g_warning("%s: %d: do_run_shell_command: couldn't fork\n", __FILE__, __LINE__);
	}

	/* Note that the forked processes might be scheduled by the operating
	 * system 'out of order', if we've made rapid successive calls to this
	 * routine.  So we try to ensure in-order execution by trying to let
	 * the child process at least start running.  And we can do this by
	 * yielding our time-slice ... */
	sched_yield();
}

void
run_shell_command (GttProject *proj, gboolean do_start)
{
	char *cmd;
	const char *str;

	cmd = (do_start) ? config_shell_start : config_shell_stop;

	if (!cmd) return;

	/* Sometimes, we are called to stop a NULL project.
	 * We don't really want that (its a result be being called twice).
	 */
	if (!proj) return;

	str = printf_project (cmd, proj);
	 do_run_shell_command (str);
	g_free ((gchar *) str);
}

/* ============================================================= */

void
cur_proj_set (GttProject *proj)
{
	/* Due to the way the widget callbacks work,
	 * we may be called recursively ... */
	if (cur_proj == proj) return;

	log_proj(NULL);
	gtt_project_timer_stop (cur_proj);
	gtt_status_icon_stop_timer (proj);
	run_shell_command (cur_proj, FALSE);

	GttProject *old_prj = cur_proj;
	if (proj) 
	{
		if (timer_is_running ())
		{
			stop_main_timer ();
		}
		cur_proj = proj;
		gtt_project_timer_start (proj);
		gtt_status_icon_start_timer (proj);
		run_shell_command (cur_proj, TRUE);
		start_idle_timer ();
		start_main_timer ();
	}
	else
	{
		if (timer_is_running ())
		{
			stop_main_timer ();
		}
		cur_proj = NULL;
		start_no_project_timer ();
	}
	log_proj(proj);

	if (old_prj)
	{
		gtt_projects_tree_update_project_data (projects_tree, old_prj);
	}

	if (cur_proj)
	{
		gtt_projects_tree_update_project_data (projects_tree, cur_proj);
	}

	/* update GUI elements */
	menu_set_states();
	toolbar_set_states();
	if (proj)
	{
		prop_dialog_set_project(proj);
		notes_area_set_project (global_na, proj);
	}
	update_status_bar();
}


/* ============================================================= */

void
focus_row_set (GttProject *proj)
{
	/* update GUI elements */

	prop_dialog_set_project(proj);
	notes_area_set_project (global_na, proj);
}


/* ============================================================= */

void
app_new(int argc, char *argv[], const char *geometry_string)
{
	GtkWidget *vbox;
	GtkWidget *widget;
	GtkWidget *vpane;
	GtkWidget *separator;
	GtkLabel *filler;
	GtkHBox *labels;
	GtkVBox *status_vbox;
	GtkStatusbar *grip;

	app_window = gnome_app_new(GTT_APP_NAME, GTT_APP_TITLE " " VERSION);
	gtk_window_set_wmclass(GTK_WINDOW(app_window),
	                         GTT_APP_NAME, GTT_APP_PROPER_NAME);

	/* 485 x 272 seems to be a good size to default to */
	gtk_window_set_default_size(GTK_WINDOW(app_window), 485, 272);
	gtk_window_set_resizable (GTK_WINDOW(app_window), TRUE);

	/* build menus */
	menus_create(GNOME_APP(app_window));

	/* build toolbar */
	widget = build_toolbar();
	gtk_widget_show(widget);
	gnome_app_set_toolbar(GNOME_APP(app_window), GTK_TOOLBAR(widget));

	/* container holds status bar, main ctree widget */
	vbox = gtk_vbox_new(FALSE, 0);

	/* build statusbar */

	status_vbox = GTK_VBOX(gtk_vbox_new(FALSE, 0));
	gtk_widget_show(GTK_WIDGET(status_vbox));

	labels = GTK_HBOX(gtk_hbox_new(FALSE, 0));
	gtk_widget_show(GTK_WIDGET(labels));

	status_bar = gtk_hbox_new(FALSE, 0);
	gtk_widget_show(status_bar);
	separator = gtk_hseparator_new();
	gtk_widget_show(separator);
	gtk_box_pack_start(GTK_BOX(status_vbox), separator, FALSE, FALSE, 0);
	gtk_box_pack_start(GTK_BOX(status_vbox), GTK_WIDGET (labels), TRUE, TRUE, 0);
	gtk_box_pack_start(GTK_BOX(status_bar), GTK_WIDGET (status_vbox), TRUE, TRUE, 0);

	grip = GTK_STATUSBAR(gtk_statusbar_new());
	gtk_statusbar_set_has_resize_grip(grip, TRUE);
	gtk_widget_show(GTK_WIDGET(grip));
	gtk_box_pack_start(GTK_BOX(status_bar), GTK_WIDGET(grip), FALSE, FALSE, 0);

	/* put elapsed time into statusbar */
	status_day_time = GTK_LABEL(gtk_label_new(_("00:00")));
	gtk_widget_show(GTK_WIDGET(status_day_time));

	gtk_box_pack_start(GTK_BOX(labels), GTK_WIDGET(status_day_time),
	                     FALSE, TRUE, 0);

	/* put project name into statusbar */
	status_project = GTK_LABEL(gtk_label_new( _("Timer is not running")));
	gtk_widget_show(GTK_WIDGET(status_project));

	gtk_box_pack_start(GTK_BOX(labels), GTK_WIDGET(status_project),
	                     FALSE, TRUE, 10);

	filler = GTK_LABEL(gtk_label_new(""));
	gtk_widget_show(GTK_WIDGET(filler));
	gtk_box_pack_start(GTK_BOX(labels), GTK_WIDGET(filler), TRUE, TRUE, 1);

	/* put timer icon into statusbar */
	status_timer = gtk_image_new_from_stock (GNOME_STOCK_TIMER,
	                                           GTK_ICON_SIZE_MENU);
	gtk_widget_show(status_timer);
	gtk_box_pack_end(GTK_BOX(status_bar), GTK_WIDGET(status_timer),
	                   FALSE, FALSE, 1);

	/* create the main columned tree for showing projects */
	projects_tree = gtt_projects_tree_new ();

	g_signal_connect (projects_tree, "columns-setup-done", G_CALLBACK (projects_tree_columns_setup_done), NULL);

	gtk_tree_view_set_reorderable (GTK_TREE_VIEW (projects_tree), TRUE);

	g_signal_connect (projects_tree, "row-activated", G_CALLBACK (projects_tree_row_activated), NULL);

	/* create the notes area */
	global_na = notes_area_new();
	vpane = notes_area_get_widget (global_na);

	/* Need to reparent, to get rid of glade parent-window hack.
	 * But gtk_widget_reparent (vpane); causes  a "Gtk-CRITICAL"
	 * to occur.  So we need a fancier move.
	 */
	gtk_widget_ref (vpane);
	gtk_container_remove(GTK_CONTAINER(vpane->parent), vpane);
	gtk_box_pack_start(GTK_BOX(vbox), vpane, TRUE, TRUE, 0);
	gtk_widget_unref (vpane);

	gtk_box_pack_end(GTK_BOX(vbox), status_bar, FALSE, FALSE, 2);

	notes_area_add_projects_tree (global_na, projects_tree);

	/* we are done building it, make it visible */
	gtk_widget_show(vbox);
	gnome_app_set_contents(GNOME_APP(app_window), vbox);

	gtt_status_icon_create();
	if (!geometry_string) return;

	if (gtk_window_parse_geometry(GTK_WINDOW(app_window),geometry_string))
	{
		geom_size_override=TRUE;
	}
	else
	{
		gnome_app_error(GNOME_APP(app_window),
			_("Couldn't understand geometry (position and size)\n"
			  " specified on command line"));
	}
}

void
app_show (void)
{
	if (!GTK_WIDGET_MAPPED(app_window))
	{
		gtk_widget_show(app_window);
	}
}



void
app_quit(GtkWidget *w, gpointer data)
{
	save_properties ();
	save_projects ();
	gtt_status_icon_destroy ();
	gtk_main_quit();
}

/* ============================== END OF FILE ===================== */
